Rich media support

hamchapman opened the pull request.

hamchapman commented:
Still needs some cleanup.

Also, @hot-leaf-juice I'm getting a `NullPointerException` here:

```java
final int position = user.getCursors().get(roomId).getPosition();
```

Any ideas why? It basically says that I'm calling `getPosition()` on `null`

hot-leaf-juice commented:
@hamchapman I suspect that there is no cursor set for `roomId`, so the `get` returns `null`. Where is that code? If that's in the library somewhere I should fix it to do a null check, otherwise I guess it's fair for us to leave that to the consumer?

vivangkumar commented on chatkit/src/main/java/com/pusher/chatkit/CurrentUser.kt:
```diff
@@ -173,34 +183,114 @@ class CurrentUser(
+    ): Cancelable? {
+        var attachmentBody: AttachmentBody? = null
+
+        if (attachment != null) {
```
We can use some async await magic here if you want to clean it up. Not that I am against this, but I reverted to this just to be able to debug that error we had.

hamchapman commented:
Now I have at least the smallest grasp of how the coroutines stuff is supposed to work, it's actually pretty sweet.

Updated a couple of the functions in the code paths I had updated:

https://github.com/pusher/chatkit-android/pull/4/files#diff-e6bbe2d672f6bb996064847db66d2b7cR279
https://github.com/pusher/chatkit-android/pull/4/files#diff-e6bbe2d672f6bb996064847db66d2b7cR235

Also, this part of it still needs to be sorted (i.e. I have no idea what I'm doing with snapshots):

https://github.com/pusher/chatkit-android/pull/4/files#diff-1bdb8e35c53c9e3cf971ea14eb88b101R42

hamchapman commented:
@vivangkumar and @hot-leaf-juice can you please take (another) look at it now that I've made the coroutine changes? 👍

hot-leaf-juice commented on app/src/main/java/com/pusher/chatkit/sample/ChatRoomActivity.java:
```diff
@@ -128,38 +161,194 @@ public void onError(Error error) {
+    }
 
-    private void showNewMessageDialog() {
+    public final void throwUpThePicker() {
```
🤮

vivangkumar commented on chatkit/build.gradle:
```diff
@@ -33,6 +38,8 @@ dependencies {
     implementation "org.jetbrains.kotlin:kotlin-stdlib-jre7:$kotlin_version"
-    api 'com.pusher:pusher-platform:0.0.1-SNAPSHOT' //Remote pusher platform
     api 'org.jetbrains.kotlinx:kotlinx-coroutines-core:0.21.2'
+//    api 'com.pusher:pusher-platform:0.0.1-SNAPSHOT' //Remote pusher platform
```
This bit really just needs to be uncommented and the line below removed. But we have to make sure we release the platform lib first and update the snapshot.

vivangkumar commented on settings.gradle:
```diff
@@ -1,5 +1,3 @@
-
-
-
+include ':app', ':chatkit', ':pusher-platform-local'
```
also all of this needs to go

hot-leaf-juice commented on chatkit/src/main/java/com/pusher/chatkit/Result.kt:
```diff
@@ -0,0 +1,34 @@
+package com.pusher.chatkit
+
+/**
+ * Code from https://github.com/michaelbull/kotlin-result
```
Can we include import this with gradle a al https://github.com/michaelbull/kotlin-result#installation?

hamchapman replied:
Could do but that has a load more utility functions for interacting with the `Result` that we don't need.

hot-leaf-juice replied:
fair enough. I'm a little bit fuzzy on whether a link to the lisence is enough to fulfil

> provided that the above copyright notice and this permission notice appear in all copies

but I'm sure I'm splitting hairs here 😛
